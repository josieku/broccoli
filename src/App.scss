$cream-white: #EDE8E8;
$broccoli-green: #5E8433;
$active-green: #4C6A29;
.App {
    position: relative;
    background: $cream-white;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    min-width: 300px;
    font-family: SpaceMono;
    height: 100%;
    overflow: hidden;
}

.LoadForeground {
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 5;
    opacity: 1;
    &.loading {
        display: block;
        background: $broccoli-green;
    }
    &.fade {
        display: block;
        background: $broccoli-green;
        animation: fadeOut 1000ms;
    }
    display: none;
}

.App-header {
    height: calc(15vh - 1px);
    min-height: 75px;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    flex-grow: 1;
    border-bottom: 1px solid black;
    padding: 0 5% 0 5%;
    div.Logo,
    div.MenuButton {
        height: 100%;
        display: flex;
        flex-direction: row;
        align-items: center;
    }
    .Logo {
        h1 {
            font-size: 1.75em;
            cursor: pointer;
            user-select: none;
        }
        img {
            width: 4em;
            height: auto;
            margin: 0 0.5em 0 0;
            transform: rotate(-20deg);
        }
        &:hover {
            img {
                animation: shake 0.82s cubic-bezier(.36, .07, .19, .97) both;
                animation-iteration-count: 2;
                backface-visibility: hidden;
                perspective: 1000px;
            }
        }
    }
}

.App-content {
    height: 70vh;
    min-height: 315px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    text-align: center;
    h1 {
        // font-size: calc(22px + 2vmin);
        font-size: 2.5em;
        margin: 0;
    }
    p {
        font-size: 1.25em;
        margin: 1.5em 0 1.5em 0;
    }
    button {
        background: $broccoli-green;
        border: 1px solid black;
        outline: none;
        padding: 1em;
        font-family: SpaceMono;
        cursor: pointer;
        color: white;
        font-size: 1.25em;
    }
    button:hover {
        background: $active-green;
    }
    button:active {
        background: $active-green;
        border: 1px solid black;
    }
}

.App-footer {
    border-top: 1px solid black;
    height: calc(15vh - 1px);
    min-height: 75px;
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    text-align: center;
    p {
        margin: 0 0 0.5em 0;
        font-size: 0.75em;
        font-style: italic;
    }
}

// HEADER/FOOTER PORPORTIONS BASED ON HEIGHT
@media (min-height: 800px) {
    .App-header,
    .App-footer {
        height: calc(12vh - 1px);
    }
    .App-content {
        height: 76vh;
    }
}

// FONT SIZES BASED ON WIDTH
@media (min-width: 450px) and (max-width: 600px) {
    .App-header .Logo img {
        width: 3em;
        // height: 45%;
    }
    .App-header .Logo h1 {
        font-size: 1.5em;
    }
    .App-content h1 {
        font-size: 1.75em;
    }
    .App-content p,
    .App-content button {
        font-size: 0.75em;
    }
}

@media (min-width: 280px) and (max-width: 449px) {
    .App-header .Logo img {
        width: 2em;
    }
    .App-header .Logo h1 {
        font-size: 1em;
    }
    .App-content h1 {
        font-size: 1.5em;
    }
    .App-content p,
    .App-content button {
        font-size: 0.8em;
    }
    .App-footer p {
        font-size: 0.6em;
    }
}

@media (max-width: 279px) {
    .App-header .Logo img {
        width: 1.5em;
    }
    .App-header .Logo h1 {
        font-size: 0.9em;
    }
    .App-content h1 {
        font-size: 1.4em;
    }
    .App-content p,
    .App-content button {
        font-size: 0.75em;
    }
    .App-footer p {
        font-size: 0.6em;
    }
}

@keyframes shake {
    10%,
    90% {
        transform: rotate(-15deg);
    }
    20%,
    80% {
        transform: rotate(0deg);
    }
    30%,
    50%,
    70% {
        transform: rotate(-15deg);
    }
    40%,
    60% {
        transform: rotate(-30deg);
    }
}

@keyframes fadeOut {
    0% {
        opacity: 1;
    }
    100% {
        opacity: 0;
    }
}