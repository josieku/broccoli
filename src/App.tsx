import { useState, useEffect } from 'react';
import Modal from './Modal';
import Menu from './Menu';
import MenuButton from './MenuButton';
import Logo from './assets/broccoli.png';
import './App.scss';

function App() {
  // loading state: "loading" (green), "fade" (fadeout animation), "" (loaded)
  const [isLoading, setIsLoading] = useState("loading");
  
  // menu & modal states: "" (closed), "close" (closing animation), "open" (opened)
  const [menuShow, setMenuShow] = useState(""); 
  const [modalShow, setModalShow] = useState("");

  useEffect(() => { // set loading states on component mount
    setTimeout(()=>setIsLoading("fade"), 1000); // start fading after 1s
    setTimeout(()=>setIsLoading(""), 2000); // clear load screen after 2s
  }, [])

  const closeModal = () => {
    setModalShow("close");
    setTimeout(()=>setModalShow(""), 400);
  }

  const closeMenu = () => {
    setMenuShow("close");
    setTimeout(()=>setMenuShow(""), 700);
  }

  return (
    <div className="App">
      <div className={`LoadForeground ${isLoading}`}/>
      <header className="App-header">
        <div className="Logo">
          <img src={Logo} alt="Logo"/>
          <h1>BROCCOLI {"&"} CO.</h1>
        </div>
        <MenuButton open={()=>setMenuShow("open")}/>
      </header>
      <div className="App-content">
        <h1>A better way</h1>
        <h1>to enjoy every day.</h1>
        <p>Be the first to know when we launch.</p>
        <button onClick={()=>setModalShow("open")}>Request an invite</button>
      </div>
      <footer className="App-footer">
        <p>Made with ðŸ’š in Shanghai</p>
        <p>Â© 2020 Broccoli {"&"} Co. All rights reserved.</p>
      </footer>
      <Modal show={modalShow} onClose={closeModal}/>
      <Menu show={menuShow} onClose={closeMenu}/>
    </div>
  );
}

export default App;
