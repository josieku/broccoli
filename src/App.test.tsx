import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { shallow, mount } from 'enzyme';
import App from './App';

test('renders the component', () => {
  const component = shallow(<App />);
  expect(component).toMatchSnapshot();
});

test("loading screen at start", () => {
  const component = shallow(<App/>);
  expect(component.find('.LoadForeground').hasClass('loading')).toEqual(true);
})

test("loading screen goes away after 2s", () => {
  jest.useFakeTimers();
  const component = mount(<App/>);
  act(()=>{
    jest.runAllTimers();
    component.update();
    expect(component.find('.LoadForeground').hasClass('loading')).toEqual(false);
    expect(component.find('.LoadForeground').hasClass('fade')).toEqual(false);
  })
})

describe('after loading screen', () => {
  let component: any;
  beforeEach(()=>{
    jest.useFakeTimers();
    component = mount(<App/>);
    act(()=>{
      jest.runAllTimers();
      component.update();
    })
  })

  test('modal is not visible at start', async () => {
    expect(component.find('Modal').getDOMNode()).not.toBeVisible();
  })
  
  test('modal is visible on click', () => {
    expect(component.find('Modal').at(0).getDOMNode()).not.toBeVisible();
    component.find('.App-content button').simulate('click');
    expect(component.find('Modal').getDOMNode()).toBeVisible();
  })

  test('modal closes on click background click', () => {
    component.find('.App-content button').simulate('click');
    component.find('.Modal-container').simulate('click');
    act(() => {
      jest.runAllTimers(); // wait for closing animation to finish
      component.update();
      expect(component.find('Modal').at(0).getDOMNode()).not.toBeVisible();
    })
  })

  test('modal closes on OK button click', async () => {
    component.find('.App-content button').simulate('click'); // open modal
    
    // fill in input fields
    component.find('input').at(0).simulate("change", { target: { value: 'test' }});
    component.find('input').at(1).simulate("change", { target: { value: 'test@test.com' }});
    component.find('input').at(2).simulate("change", { target: { value: 'test@test.com' }});

    // mock successful fetch
    const mockSuccessFetchPromise: Promise<any> = Promise.resolve({
      json: () => Promise.resolve("Registered"),
    });
    jest.spyOn(global, 'fetch').mockImplementation(():Promise<any> => mockSuccessFetchPromise);
    expect(component.find('.Modal-box.success').length).toBe(0);
    await act(async () => {
      jest.useRealTimers();
      component.find('.Modal-box button.box').simulate('click');
      await new Promise(resolve => setTimeout(resolve));
      component.update();
    });

    expect(component.find('.Modal-box.success').length).toBe(1); // find success modal 

    component.find('.Modal-box.success button.box').simulate('click'); // close modal
    expect(component.find('Modal').at(0).hasClass('open')).toEqual(false);
  })
  
  test('side bar is visible on click, by checking classname', () => {
    expect(component.find('.Side-menu').hasClass('open')).toEqual(false);
    component.find('.MenuButton').simulate('click');
    expect(component.find('.Side-menu').hasClass('open')).toEqual(true);
  })

  test('side bar closes by click on background', () => {
    component.find('.MenuButton').simulate('click');
    component.find('.Side-menu .background').simulate('click');
    act(() => {
      jest.runAllTimers(); // wait for closing animation to finish
      component.update();
      expect(component.find('.Side-menu').hasClass('open')).toEqual(false);
    })
  })
})