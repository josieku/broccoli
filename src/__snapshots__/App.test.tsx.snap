// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  className="App"
>
  <div
    className="LoadForeground loading"
  />
  <header
    className="App-header"
  >
    <div
      className="Logo"
    >
      <img
        alt="Logo"
        src="broccoli.png"
      />
      <h1>
        BROCCOLI 
        &
         CO.
      </h1>
    </div>
    <MenuButton
      open={[Function]}
    />
  </header>
  <div
    className="App-content"
  >
    <h1>
      A better way
    </h1>
    <h1>
      to enjoy every day.
    </h1>
    <p>
      Be the first to know when we launch.
    </p>
    <button
      onClick={[Function]}
    >
      Request an invite
    </button>
  </div>
  <footer
    className="App-footer"
  >
    <p>
      Made with ðŸ’š in Shanghai
    </p>
    <p>
      Â© 2020 Broccoli 
      &
       Co. All rights reserved.
    </p>
  </footer>
  <Modal
    onClose={[Function]}
    show=""
  />
  <Menu
    onClose={[Function]}
    show=""
  />
</div>
`;
